### ISSUE TITLE
Accessing RTTI During Decoration

### ISSUE THREAD

**Issue by dhkatz** [typescript-rtti/typescript-rtti#76]
So I'm aware of how decoration order is supposed to work with classes:

```
Parameters -> Methods -> Class
```

For example:

```typescript

@cls()
class Foo {
  @method()
  public bar(@param() baz: number): Promise<void> {}
```

Normally TypeScript will decorate `@param()` then `@method()`, followed by `@cls()` on the whole class.

However, `typescript-rtti` seems to mess with this order somewhat.

For example, I am not able to call `Reflect.getMetadata('design:paramtypes', ...)` anymore inside of `@param()` because the library decorates the design metadata late. Same with `design:type` and `design:returntype`.

```typescript
__decorate([
    (0, src_1.method)(),
    __RΦ.m("design:type", Function),
    __RΦ.m("design:returntype", Promise),
    __param(0, (0, src_1.param)()) // This should go AFTER the two decorators above, not before
], Foo.prototype, "bar", null);

Foo= __decorate([
    (0, src_1.cls)()
], Foo);

// Then later on...

__RΦ.m("rt:p", [{ n: "baz", t: () => __RΦ.a(16), v: null }])(Foo.prototype, "bar");
__RΦ.m("design:paramtypes", [Number])(Foo.prototype, "bar");
```

I *don't* expect to be able to access the full metadata about a class/method within a parameter decorator, but I expect to at least be able to get the same information that TypeScript was providing before.

I also can't access information about methods within a class decorator, but I expect most of that information to be available at that point (because of the order of decoration I initially mentioned)

The same goes for method decorators, I expect most information about parameters to be available in a method decorator but there isn't.

In the current state of the library, I can't access much of the information until after the *entire* class has been defined, meaning I have to move a lot of logic from decorators into functions that will have to manually process classes later

**Comment by rezonant** [comment]
I think this could be fixed by detecting that there is at least one decorator applied to an element and then moving the generated metadata declaration calls into the top of the decorators array. The reason we do not emit into the decorators array by default is because TS will not emit them if there is not at least one decorator applied in the original source.

**Comment by rezonant** [comment]
I'm also not certain the original decision to hijack emitDecoratorMetadata is still relevant, as since then RTTI has been given a completely independent import generation system, so we might be able to safely enable TS' own emitDecoratorMetadata implementation without breaking the RTTI emit

**Comment by dhkatz** [comment]
The emitDecoratorMetadata comment was kinda just as an extra point that it's the only way to get even a little info inside the decorators is by enabling it. Ideally, I wouldn't want to have it on like you mentioned, because I'd rather rely on the RTTI added by the library

**Comment by rezonant** [comment]
OK so the emitting of metadata outside of `__decorate` in so many places is me being a bit sloppy. When you see a metadata decorator emit after the declaration, that is called an _outboard decorator_ in the lingo of the `typescript-rtti` codebase. Outboard decorators are required to emit metadata for syntactic elements which cannot have user decorators applied to them (class expression is a good example).

I've fixed this up quite a bit so that we only emit using outboard decorators when we actually have to, which does restore the ordering so that user decorators on classes, methods and properties can access RTTI metadata. 

However, parameter decorators are another thing. Unfortunately, due to the way that Typescript constructs the final decorators array when emitting JS, I don't think it will be possible to emit RTTI decorators in a way that makes them accessible during user decorators applied to parameters. It's worth noting that this only ever worked for the traditional emit for `design:*`, as parameter decorators always run before the method decorators, so if you were not relying on `design:*`, you couldn't accomplish this anyway.

This is the crux of the issue, in `checker.ts:1120` (or so, my copy is probably behind HEAD):

![image](https://user-images.githubusercontent.com/409140/176334665-048245a1-26b9-4f24-83ac-5ef4a8bd5004.png)

**Comment by rezonant** [comment]
This work, along with all of vNext is currently blocked by what appears to be a Typescript bug

**Comment by dhkatz** [comment]
What sort of bug is causing the issue?

**Comment by rezonant** [comment]
Yeah I asked Ryan (TS engineering lead) if I'm just doing something dumb but haven't heard back so I'm planning to file a bug

**Comment by rezonant** [comment]
This issue is blocked by https://github.com/microsoft/TypeScript/issues/49794

**Comment by dhkatz** [comment]
A shame, but hopefully it will be resolved in the future 🙂

**Comment by dhkatz** [comment]
Is there any update on this issue? I know TS has had a view minor/major versions since (and decorators changing completely in 5.0)
